### ISSUE TITLE
evp_md_init_internal: type checks are superfluous + optimizations lost compared to 1.1.1

### ISSUE THREAD

**Issue by dlegaultbbry** [openssl/openssl#16947]
In 3.0 in evp_md_init_internal, even if you specify type == NULL in the function call, you always inevitably get one in this block

```
    if (type != NULL) {
        ctx->reqdigest = type;
    } else {
        if (ctx->digest == NULL) {
            ERR_raise(ERR_LIB_EVP, EVP_R_NO_DIGEST_SET);
            return 0;
        }
        type = ctx->digest;
    }
```

There are many checks for type == NULL or type != NULL after that point which are unnecessary.

At the beginning of the function, it also forces the algctx of the provider to be free'd even if the algorithm is re-used wasting an opportunity for optimization if the algorithm is the same and comes from the same provider.

For example, this block was optional before and only done if you changed the digest type:

```
    if (ctx->digest != NULL && ctx->digest->ctx_size > 0) {
        OPENSSL_clear_free(ctx->md_data, ctx->digest->ctx_size);
        ctx->md_data = NULL;
    }
```

Now it's unconditionally executed leading to more wasted time and resources (see legacy path for much of the old 1.1.1 implementation).

There's also some ctx->engine checks which are not inside #if !defined(OPENSSL_NO_ENGINE).

**Comment by paulidale** [comment]
Partially addressed in #16952, where the unnecessary NULL checks are removed and the check of ->engine is moved inside the #if guard.

**Comment by paulidale** [comment]
This part remains outstanding:

> At the beginning of the function, it also forces the algctx of the provider to be free'd even if the algorithm is re-used wasting an opportunity for optimization if the algorithm is the same and comes from the same provider.
